File: app/assets/javascripts/cqm/PopulationSet.js
/* eslint-disable no-unused-vars, no-param-reassign */
const mongoose = require('mongoose/browser');
const { StatementReferenceSchema } = require('./CQLStatementDependency');

const [Mixed] = [
  mongoose.Schema.Types.Mixed,
];

// TODO: figure out if there is a better way to handle inheritance like mongoid.
//  _type is how mongoid stores the specific type of the embedded document.
const PopulationMapSchema = new mongoose.Schema({
  _type: String,
  IPP: StatementReferenceSchema,
  DENOM: StatementReferenceSchema,
  NUMER: StatementReferenceSchema,
  NUMEX: StatementReferenceSchema,
  DENEX: StatementReferenceSchema,
  DENEXCEP: StatementReferenceSchema,
  MSRPOPL: StatementReferenceSchema,
  MSRPOPLEX: StatementReferenceSchema,
  // STRAT is only here so cqm-execution can handle stratification results compliation with the current approach.
  STRAT: StatementReferenceSchema,
});

if (!PopulationMapSchema.options.toObject) PopulationMapSchema.options.toObject = {};
PopulationMapSchema.options.toObject.transform = function transform(doc, ret, options) {
  // remove the _id and _type of every document before returning the result
  delete ret._id;
  delete ret._type;
  return ret;
};


const StratificationSchema = new mongoose.Schema({
  title: String,
  stratification_id: String,
  hqmf_id: String,
  statement: StatementReferenceSchema,
});

const ObservationSchema = new mongoose.Schema({
  title: String,
  observation_function: StatementReferenceSchema,
  observation_parameter: StatementReferenceSchema,
  aggregation_type: String,
  hqmf_id: String,
});

const PopulationSetSchema = new mongoose.Schema({
  title: String,
  population_set_id: String,
  populations: PopulationMapSchema,
  stratifications: [StratificationSchema],
  supplemental_data_elements: [StatementReferenceSchema],
  observations: [ObservationSchema],
});

module.exports.StratificationSchema = StratificationSchema;
class Stratification extends mongoose.Document {
  constructor(object) {
    super(object, StratificationSchema);
  }
}
module.exports.Stratification = Stratification;

module.exports.ObservationSchema = ObservationSchema;
class Observation extends mongoose.Document {
  constructor(object) {
    super(object, ObservationSchema);
  }
}
module.exports.Observation = Observation;

module.exports.PopulationMapSchema = PopulationMapSchema;
class PopulationMap extends mongoose.Document {
  constructor(object) {
    super(object, PopulationMapSchema);
  }
}
module.exports.PopulationMap = PopulationMap;

module.exports.PopulationSetSchema = PopulationSetSchema;
class PopulationSet extends mongoose.Document {
  constructor(object) {
    super(object, PopulationSetSchema);
  }
}
module.exports.PopulationSet = PopulationSet;

File: app/models/cqm/population_set.rb
module CQM
  # Representation of a population set (aka. PopulationCriteria)
  class PopulationSet
    include Mongoid::Document

    embedded_in :measure

    embeds_one :populations, class_name: 'CQM::PopulationMap'
    embeds_many :stratifications
    embeds_many :supplemental_data_elements, class_name: 'CQM::StatementReference'
    embeds_many :observations

    field :title, type: String
    field :population_set_id, type: String
  end

  # Stratification info
  class Stratification
    include Mongoid::Document

    embedded_in :population_set

    field :title, type: String
    field :stratification_id, type: String
    field :hqmf_id, type: String
    embeds_one :statement, class_name: 'CQM::StatementReference'
  end

  # Observation info
  class Observation
    include Mongoid::Document

    embedded_in :population_set

    embeds_one :observation_function, class_name: 'CQM::StatementReference'
    embeds_one :observation_parameter, class_name: 'CQM::StatementReference'
    field :aggregation_type, type: String
    field :hqmf_id, type: String
  end

  # Base class for the population maps
  class PopulationMap
    include Mongoid::Document

    embedded_in :population_set
  end

  # The population map for a porportion population set
  class ProportionPopulationMap < PopulationMap
    include Mongoid::Document

    embeds_one :IPP, class_name: 'CQM::StatementReference'
    embeds_one :DENOM, class_name: 'CQM::StatementReference'
    embeds_one :NUMER, class_name: 'CQM::StatementReference'
    embeds_one :NUMEX, class_name: 'CQM::StatementReference'
    embeds_one :DENEX, class_name: 'CQM::StatementReference'
    embeds_one :DENEXCEP, class_name: 'CQM::StatementReference'
  end

  # The population map for a ratio population set
  class RatioPopulationMap < PopulationMap
    include Mongoid::Document

    embeds_one :IPP, class_name: 'CQM::StatementReference'
    embeds_one :DENOM, class_name: 'CQM::StatementReference'
    embeds_one :NUMER, class_name: 'CQM::StatementReference'
    embeds_one :NUMEX, class_name: 'CQM::StatementReference'
    embeds_one :DENEX, class_name: 'CQM::StatementReference'
  end

  # The population map for a continuous variable population set
  class ContinuousVariablePopulationMap < PopulationMap
    include Mongoid::Document

    embeds_one :IPP, class_name: 'CQM::StatementReference'
    embeds_one :MSRPOPL, class_name: 'CQM::StatementReference'
    embeds_one :MSRPOPLEX, class_name: 'CQM::StatementReference'
  end

  # The population map for a cohort population set
  class CohortPopulationMap < PopulationMap
    include Mongoid::Document

    embeds_one :IPP, class_name: 'CQM::StatementReference'
  end
end
